scheduler/feasible.go

```diff
@@ -137,87 +139,115 @@ func NewRandomIterator(ctx Context, nodes []*structs.Node) *StaticIterator {
+			h.volumeReqs = append(h.volumeReqs, &allocVolumeRequest{allocID, copied})
+
 		} else {
-			lookupMap[req.Source] = append(lookupMap[req.Source], req)
+			h.volumeReqs = append(h.volumeReqs, &allocVolumeRequest{allocID, req})
 		}
 	}
-	h.volumes = lookupMap
 }
 
 func (h *HostVolumeChecker) Feasible(candidate *structs.Node) bool {
-	if h.hasVolumes(candidate) {
+	feasible, failure := h.hasVolumes(candidate)
+	if feasible {
 		return true
 	}
 
-	h.ctx.Metrics().FilterNode(candidate, FilterConstraintHostVolumes)
+	h.ctx.Metrics().FilterNode(candidate, failure)
 	return false
 }
 
-func (h *HostVolumeChecker) hasVolumes(n *structs.Node) bool {
-	rLen := len(h.volumes)
-	hLen := len(n.HostVolumes)
-
+func (h *HostVolumeChecker) hasVolumes(n *structs.Node) (bool, string) {
 	// Fast path: Requested no volumes. No need to check further.
-	if rLen == 0 {
-		return true
-	}
-
-	// Fast path: Requesting more volumes than the node has, can't meet the criteria.
-	if rLen > hLen {
-		return false
+	if len(h.volumeReqs) == 0 {
+		return true, ""
 	}
 
-	for source, requests := range h.volumes {
-		nodeVolume, ok := n.HostVolumes[source]
+	ws := memdb.NewWatchSet()
+	for _, req := range h.volumeReqs {
+		volCfg, ok := n.HostVolumes[req.volumeReq.Source]
 		if !ok {
-			return false
+			return false, FilterConstraintHostVolumes
 		}
 
-		// If the volume supports being mounted as ReadWrite, we do not need to
-		// do further validation for readonly placement.
-		if !nodeVolume.ReadOnly {
-			continue
-		}
+		if volCfg.ID != "" { // dynamic host volume
+			vol, err := h.ctx.State().HostVolumeByID(nil, h.namespace, volCfg.ID, false)
+			if err != nil || vol == nil {
+				// node fingerprint has a dynamic volume that's no longer in the
+				// state store; this is only possible if the batched fingerprint
+				// update from a delete RPC is written before the delete RPC's
+				// raft entry completes
+				return false, FilterConstraintHostVolumes
+			}
+			if vol.State != structs.HostVolumeStateReady {
+				return false, FilterConstraintHostVolumes
+			}
+			var capOk bool
+			for _, cap := range vol.RequestedCapabilities {
+				if req.volumeReq.AccessMode == structs.CSIVolumeAccessMode(cap.AccessMode) &&
+					req.volumeReq.AttachmentMode == structs.CSIVolumeAttachmentMode(cap.AttachmentMode) {
+					capOk = true
+					break
+				}
+			}
+			if !capOk {
+				return false, FilterConstraintHostVolumes
+			}
 
-		// The Volume can only be mounted ReadOnly, validate that no requests for
-		// it are ReadWrite.
-		for _, req := range requests {
-			if !req.ReadOnly {
-				return false
+			if req.volumeReq.Sticky {
+				allocation, err := h.ctx.State().AllocByID(ws, req.allocID)
+				if err != nil {
+					return false, FilterConstraintHostVolumesAllocLookupFailed
+				}
+				if slices.Contains(allocation.HostVolumeIDs, vol.ID) || len(allocation.HostVolumeIDs) == 0 {
+					return true, ""
+				}
+
+				return false, FilterConstraintHostVolumes
+			}
```

Comment:
Rather than going back to the state store, let's add the `Allocation.HostVolumeIDs` to the [`SelectOptions`](https://github.com/hashicorp/nomad/blob/v1.9.3/scheduler/stack.go#L37-L42) struct. We can populate that value [here](https://github.com/hashicorp/nomad/blob/v1.9.3/scheduler/generic_sched.go#L827) in `getSelectOptions`, where we have access to the host. Add a `[]string` parameter for the host volume IDs to `SetVolumes` instead of the alloc ID, and then you can just do a `slices.Contains` check here on that value.

That _should_ get us what we want in terms of propagating the value from the previous allocation.